abci/proposals/validate.go

```diff
@@ -28,33 +32,80 @@ func (h *ProposalHandler) ValidateExtendedCommitInfo(
 
 	// Validate all oracle vote extensions.
 	for _, vote := range extendedCommitInfo.Votes {
-		address := sdk.ConsAddress{}
-		if err := address.Unmarshal(vote.Validator.Address); err != nil {
+		address := sdk.ConsAddress(vote.Validator.Address)
+		// The vote extension are from the previous block.
+		if err := validateVoteExtension(ctx, vote, h.voteExtensionCodec, h.currencyPairStrategy); err != nil {
 			h.logger.Error(
-				"failed to unmarshal validator address",
+				"failed to validate oracle vote extension",
 				"height", height,
+				"validator", address.String(),
+				"err", err,
 			)
 
 			return err
 		}
+	}
 
-		voteExt, err := h.voteExtensionCodec.Decode(vote.VoteExtension)
-		if err != nil {
-			return err
-		}
+	return nil
+}
 
-		// The vote extension are from the previous block.
-		if err := ve.ValidateOracleVoteExtension(ctx, voteExt, h.currencyPairStrategy); err != nil {
-			h.logger.Error(
-				"failed to validate oracle vote extension",
-				"height", height,
-				"validator", address.String(),
+// PruneAndValidateExtendedCommitInfo validates each vote-extension in the extended commit, and removes
+// any vote-extensions that are invalid. Removal will effectively treat the validator's
+// vote as absent.  This function performs all validation that ValidateExtendedCommitInfo performs.
+func (h *ProposalHandler) PruneAndValidateExtendedCommitInfo(
+	ctx sdk.Context, extendedCommitInfo cometabci.ExtendedCommitInfo,
+) (cometabci.ExtendedCommitInfo, error) {
+	// Validate all oracle vote extensions.
+	for i, vote := range extendedCommitInfo.Votes {
+		// validate the vote-extension
+		if err := validateVoteExtension(ctx, vote, h.voteExtensionCodec, h.currencyPairStrategy); err != nil {
+			h.logger.Info(
+				"failed to validate vote extension - pruning vote",
 				"err", err,
+				"validator", vote.Validator.Address,
 			)
 
-			return err
+			// failed to validate this vote-extension, mark it as absent in the original commit
+			vote.BlockIdFlag = cometproto.BlockIDFlagAbsent
+			vote.ExtensionSignature = nil
+			vote.VoteExtension = nil
+			extendedCommitInfo.Votes[i] = vote
 		}
 	}
 
+	// validate after pruning
+	if err := h.validateVoteExtensionsFn(ctx, extendedCommitInfo); err != nil {
+		h.logger.Error(
+			"failed to validate vote extensions; vote extensions may not comprise a super-majority",
+			"err", err,
+		)
+
+		return cometabci.ExtendedCommitInfo{}, err
+	}
+
+	return extendedCommitInfo, nil
+}
+
+func validateVoteExtension(
+	ctx sdk.Context,
+	vote cometabci.ExtendedVoteInfo,
+	voteExtensionCodec codec.VoteExtensionCodec,
+	currencyPairStrategy currencypair.CurrencyPairStrategy,
+) error {
+	// vote is not voted for if VE is nil
```

Comment:
wym

Reply:
why not j skip non-commit votes above?

Reply:
you mean before this function is called?  So that if a `nil` VE is passed into this function it will be considered invalid?

Reply:
my thinking is that a `nil` vote IS valid as long as there is no signature.  It is valid behavior to have this kind of vote, so to me it makes sense to live in here no?

Reply:
kk in that case, let's adjust this check to check that there shld be extensions for commit votes, and shld not for non-commit votes? Also the godoc on L95 is a bit unclear